clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -127,6 +128,24 @@ public void awaitMetadataUpdate() {
     }
 
     /**
+     * Block until the metadata has been refreshed or the timeout has expired.
+     * @param timeout The maximum duration (in ms) to wait for the request
+     * @throws TimeoutException if the timeout has expired
+     */
+    public void awaitMetadataUpdate(long timeout) {
+        long now = time.milliseconds();
+        long deadline = now + timeout;
+        int version = this.metadata.requestUpdate();
+        while (this.metadata.version() == version && now < deadline) {
```

Comment:
Yes, good points @hachikuji and @onurkaraman. By the way, I totally understand that this is a WIP. I just thought it would be worth discussing this sooner rather than later. :)

It does seem that a variant that lets the caller decide what to do would be good to have. Not sure if a boolean is good enough to represent the 3 possible states (or if the callers need to know about the 3 possible states).